sdks/java/testing/expansion-service/src/test/java/org/apache/beam/sdk/expansion/TestExpansionService.java

```diff
@@ -55,39 +88,154 @@
+                  return input
+                      .apply(
+                          FileIO.<GenericRecord>write()
+                              .via(ParquetIO.sink(schema))
+                              .to(spec.getPayload().toStringUtf8()))
+                      .getPerDestinationOutputFilenames()
+                      .apply(Values.create());
+                }
+              });
+
+      builder.put(TEST_COUNT_URN, spec -> Count.perElement());
+      builder.put(
           TEST_FILTER_URN,
-              spec ->
-                  Filter.lessThanEq(
-                      // TODO(BEAM-6587): Use strings directly rather than longs.
-                      (long) spec.getPayload().toStringUtf8().charAt(0)),
+          spec ->
+              Filter.lessThanEq(
+                  // TODO(BEAM-6587): Use strings directly rather than longs.
+                  (long) spec.getPayload().toStringUtf8().charAt(0)));
+      builder.put(
           TEST_PARQUET_READ_URN,
-              spec ->
-                  new PTransform<PBegin, PCollection<GenericRecord>>() {
-                    @Override
-                    public PCollection<GenericRecord> expand(PBegin input) {
-                      return input
-                          .apply(FileIO.match().filepattern(spec.getPayload().toStringUtf8()))
-                          .apply(FileIO.readMatches())
-                          .apply(ParquetIO.readFiles(schema))
-                          .setCoder(AvroCoder.of(schema));
-                    }
-                  },
-          TEST_PARQUET_WRITE_URN,
-              spec ->
-                  new PTransform<PCollection<GenericRecord>, PCollection<String>>() {
-                    @Override
-                    public PCollection<String> expand(PCollection<GenericRecord> input) {
-                      return input
-                          .apply(
-                              FileIO.<GenericRecord>write()
-                                  .via(ParquetIO.sink(schema))
-                                  .to(spec.getPayload().toStringUtf8()))
-                          .getPerDestinationOutputFilenames()
-                          .apply(Values.create());
-                    }
-                  });
+          spec ->
+              new PTransform<PBegin, PCollection<GenericRecord>>() {
+                @Override
+                public PCollection<GenericRecord> expand(PBegin input) {
+                  return input
+                      .apply(FileIO.match().filepattern(spec.getPayload().toStringUtf8()))
+                      .apply(FileIO.readMatches())
+                      .apply(ParquetIO.readFiles(schema))
+                      .setCoder(AvroCoder.of(schema));
+                }
+              });
+      return builder.build();
+    }
+
+    public static class TestFlattenTransformProvider
+        implements ExpansionService.TransformProvider<PCollectionList<Long>, PCollection<Long>> {
```

Comment:
Instead of implementing TransformProviders can we implement TransformRegistrar and TransformBuilder combination which is our public API for transform authors ? I think using low level API here will make this confusing for users. I think we should ditch using TestExpansionService altogether and try to implement integration tests using the real ExpansionService.

Reply:
Update to use `ExternalTransformRegistrar` whenever possible.